dronegen/container_images.go

```diff
@@ -0,0 +1,942 @@
+	}
+
+	for _, product := range rv.getProducts(clonedRepoPath) {
+		steps = append(steps, product.buildSteps(rv, setupStepNames, flags)...)
+	}
+
+	return steps
+}
+
+type semver struct {
+	Name        string // Human-readable name for the information contained in the semver, i.e. "major"
+	FilePath    string // The path under the working dir where the information can be read from
+	FieldCount  int    // The number of significant version fields available in the semver i.e. "v10" -> 1
+	IsImmutable bool
+}
+
+func (rv *releaseVersion) getSemvers() []*semver {
+	varDirectory := "/go/var"
+	return []*semver{
+		{
+			Name:        "major",
+			FilePath:    path.Join(varDirectory, "major-version"),
+			FieldCount:  1,
+			IsImmutable: false,
+		},
+		{
+			Name:        "minor",
+			FilePath:    path.Join(varDirectory, "minor-version"),
+			FieldCount:  2,
+			IsImmutable: false,
+		},
+		{
+			Name:        "canonical",
+			FilePath:    path.Join(varDirectory, "canonical-version"),
+			FieldCount:  3,
+			IsImmutable: true,
+		},
+	}
+}
+
+func (rv *releaseVersion) buildSplitSemverSteps() step {
+	semvers := rv.getSemvers()
+
+	commands := make([]string, 0, len(semvers))
+	for _, semver := range semvers {
+		// Ex: semver.FieldCount = 3, cutFieldString = "1,2,3"
+		cutFieldStrings := make([]string, 0, semver.FieldCount)
+		for i := 1; i <= semver.FieldCount; i++ {
+			cutFieldStrings = append(cutFieldStrings, strconv.Itoa(i))
+		}
+		cutFieldString := strings.Join(cutFieldStrings, ",")
+
+		commands = append(commands,
+			fmt.Sprintf("echo %s | sed 's/v//' | cut -d'.' -f %q > %q", rv.ShellVersion, cutFieldString, semver.FilePath),
+		)
+	}
+
+	return step{
+		Name:     "Build major, minor, and canonical semver",
+		Image:    "alpine",
+		Commands: commands,
+	}
+}
+
+func (rv *releaseVersion) getProducts(clonedRepoPath string) []*Product {
+	teleportOperatorProduct := NewTeleportOperatorProduct(clonedRepoPath)
+
+	products := make([]*Product, 0, 1)
+	products = append(products, teleportOperatorProduct)
+
+	return products
+}
+
+func (rv *releaseVersion) getTagsForVersion() []*ImageTag {
+	semvers := rv.getSemvers()
+	imageTags := make([]*ImageTag, 0, len(semvers))
+	for _, semver := range semvers {
+		imageTags = append(imageTags, &ImageTag{
+			ShellBaseValue:   fmt.Sprintf("$(cat %s)", semver.FilePath),
+			DisplayBaseValue: semver.Name,
+			IsImmutable:      semver.IsImmutable,
+		})
+	}
+
+	return imageTags
+}
+
+type Image struct {
+	Repo string
+	Name string
+	Tag  *ImageTag
+}
+
+func (i *Image) GetShellName() string {
+	repo := ""
+	if !i.IsLocalImage() {
+		// Ensure one and only one "/"
+		repo = strings.TrimSuffix(i.Repo, "/")
+		repo += "/"
+	}
+
+	return fmt.Sprintf("%s%s:%s", repo, i.Name, i.Tag.GetShellValue())
+}
+
+func (i *Image) GetDisplayName() string {
+	return fmt.Sprintf("%s:%s", i.Name, i.Tag.GetDisplayValue())
+}
+
+func (i *Image) IsLocalImage() bool {
+	return i.Repo == ""
+}
+
+type Product struct {
```

Comment:
Can you organize the code a little better? Instead of placing everything in a single giant file, I would for example move everything related to products into, say, product.go, container repo related stuff into another file, etc. Would be easier to follow when related things are grouped together. Maybe even put them all under a subpackage e.g. dronegen/multiarch or something.